public/app/features/scenes/variables/variants/query/QueryVariable.tsx

```diff
@@ -0,0 +1,184 @@
+import React from 'react';
+import { Observable, Subject, of, Unsubscribable, filter, take, mergeMap, catchError, throwError } from 'rxjs';
+import { v4 as uuidv4 } from 'uuid';
+
+import {
+  CoreApp,
+  DataQuery,
+  DataQueryRequest,
+  DataSourceApi,
+  DataSourceRef,
+  getDefaultTimeRange,
+  LoadingState,
+  PanelData,
+  ScopedVars,
+  VariableRefresh,
+  VariableSort,
+} from '@grafana/data';
+import { getDataSourceSrv } from '@grafana/runtime';
+import { toMetricFindValues } from 'app/features/variables/query/operators';
+
+import { sceneGraph } from '../../../core/sceneGraph';
+import { SceneComponentProps } from '../../../core/types';
+import { VariableDependencyConfig } from '../../VariableDependencyConfig';
+import { VariableValueSelect } from '../../components/VariableValueSelect';
+import { VariableValueOption } from '../../types';
+import { MultiValueVariable, MultiValueVariableState, VariableGetOptionsArgs } from '../MultiValueVariable';
+
+import { createQueryVariableRunner } from './createQueryVariableRunner';
+import { metricNamesToVariableValues } from './utils';
+
+export interface QueryVariableState extends MultiValueVariableState {
+  datasource: DataSourceRef | null;
+  query: any;
+  regex: string;
+  refresh: VariableRefresh;
+  sort: VariableSort;
+}
+
+export class QueryVariable extends MultiValueVariable<QueryVariableState> {
+  private updateSubscription?: Unsubscribable;
+  private dataSourceSubject?: Subject<DataSourceApi>;
+
+  protected _variableDependency = new VariableDependencyConfig(this, {
+    statePaths: ['regex'],
+    // TODO: add query and datasource support
+  });
+
+  public constructor(initialState: Partial<QueryVariableState>) {
+    super({
+      name: '',
+      value: '',
+      text: '',
+      query: '',
+      options: [],
+      datasource: null,
+      regex: '',
+      refresh: VariableRefresh.onDashboardLoad,
+      sort: VariableSort.alphabeticalAsc,
+      ...initialState,
+    });
+  }
+
+  public activate(): void {
+    super.activate();
+    const timeRange = sceneGraph.getTimeRange(this);
+
+    if (this.state.refresh === VariableRefresh.onTimeRangeChanged) {
+      this._subs.add(
+        timeRange.subscribeToState({
+          next: () => {
+            this.updateSubscription = this.validateAndUpdate().subscribe();
+          },
+        })
```

Comment:
If multiple dependent variables (A -> B -> C) depend on time range won't this start all at the same time?

Reply:
They will, I'll try to figure this out in a followup work